File: flux/flux-hardware/wasm/flux_vm_wasm.js
/**
 * FLUX-C Constraint Checker VM — JavaScript wrapper
 * Loads the WASM module and provides a friendly FluxVM class.
 * Falls back to a pure-JS interpreter if WASM is unavailable.
 */

const OP = {
  HALT: 0x1a, ASSERT: 0x1b, RANGE: 0x1d,
  BOOL_AND: 0x26, BOOL_OR: 0x27, DUP: 0x28, SWAP: 0x29,
};

/* ── Pure JS fallback interpreter ── */
function jsCheck(bytecode, input) {
  const stack = [input];
  let gas = 0;
  for (let pc = 0; pc < bytecode.length; pc++) {
    if (++gas >= 0xffff) return 2;
    const op = bytecode[pc];
    switch (op) {
      case OP.HALT: return stack.pop() === 0 ? 0 : 1;
      case OP.ASSERT: { if (!stack.pop()) return 1; break; }
      case OP.RANGE: { const h=stack.pop(),l=stack.pop(),v=stack.pop(); stack.push(v>=l&&v<=h?1:0); break; }
      case OP.BOOL_AND: { const b=stack.pop(),a=stack.pop(); stack.push(a&&b?1:0); break; }
      case OP.BOOL_OR:  { const b=stack.pop(),a=stack.pop(); stack.push(a||b?1:0); break; }
      case OP.DUP:  { stack.push(stack[stack.length-1]); break; }
      case OP.SWAP: { const t=stack[stack.length-1]; stack[stack.length-1]=stack[stack.length-2]; stack[stack.length-2]=t; break; }
      default: return 3;
    }
  }
  return stack.length && stack[0] !== 0 ? 1 : 0;
}

/* ── FluxVM class ── */
class FluxVM {
  constructor() { this._wasm = null; this._mem = null; }

  async init(wasmUrl = 'flux_vm.wasm') {
    try {
      const resp = await fetch(wasmUrl);
      const { instance } = await WebAssembly.instantiate(await resp.arrayBuffer(), {});
      this._wasm = instance.exports;
      this._mem  = this._wasm.memory;
    } catch (_) {
      this._wasm = null;  // will use JS fallback
    }
    return this;
  }

  /** Check a single value. Returns 0=pass, 1=fail, 2=gas, 3=error. */
  check(value, bytecode) {
    if (!this._wasm) return jsCheck(bytecode, value);
    const mem = new Uint8Array(this._mem.buffer);
    const bcOff = 64;                           // offset for bytecode
    mem.set(bytecode, bcOff);
    return this._wasm.flux_check(bcOff, bytecode.length, value);
  }

  /** Batch-check an array of values. Returns array of result codes. */
  batch(values, bytecode, maxGas = 0xffff) {
    if (!this._wasm) return values.map(v => jsCheck(bytecode, v));

    const mem = new Uint8Array(this._mem.buffer);
    const bcOff   = 64;
    const inOff   = bcOff + bytecode.length + 4 & ~3;   // align 4
    const outOff  = inOff + values.length * 4;
    const needed  = outOff + values.length * 4;

    if (this._mem.buffer.byteLength < needed) {
      this._mem.grow(Math.ceil((needed - this._mem.buffer.byteLength) / 65536));
    }
    const mem2 = new Uint8Array(this._mem.buffer);
    mem2.set(bytecode, bcOff);
    const inArr = new Int32Array(this._mem.buffer, inOff, values.length);
    inArr.set(values);

    this._wasm.flux_batch(bcOff, bytecode.length, inOff, outOff, values.length, maxGas);

    const outArr = new Int32Array(this._mem.buffer, outOff, values.length);
    return Array.from(outArr);
  }
}

// ESM / CJS dual export
if (typeof module !== 'undefined') module.exports = { FluxVM };
else if (typeof globalThis !== 'undefined') globalThis.FluxVM = FluxVM;

File: flux/flux-hardware/wasm/flux_vm_wasm.c
/*
 * FLUX-C Constraint Checker VM — WebAssembly implementation
 * Compile: clang --target=wasm32-unknown-none -O2 -nostdlib -Wl,--no-entry \
 *          -Wl,--export=flux_init -Wl,--export=flux_check -Wl,--export=flux_batch \
 *          -Wl,--export=__heap_base flux_vm_wasm.c -o flux_vm.wasm
 *
 * Alternatively: wasm-ld or emscripten with -s STANDALONE_WASM
 */

#define STACK_SIZE 64
#define MAX_GAS    0xFFFF

/* Opcodes */
#define OP_HALT     0x1A
#define OP_ASSERT   0x1B
#define OP_RANGE    0x1D
#define OP_BOOL_AND 0x26
#define OP_BOOL_OR  0x27
#define OP_DUP      0x28
#define OP_SWAP     0x29

typedef signed int     i32;
typedef unsigned int   u32;
typedef unsigned char  u8;
typedef unsigned short u16;

/* VM state — lives in linear memory */
static i32 stack[STACK_SIZE];
static i32 sp;           /* stack pointer (next free slot) */
static u16 gas;

/* ── helpers (static inline so no external deps) ── */

static inline void push(i32 v) {
    if (sp < STACK_SIZE) stack[sp++] = v;
}

static inline i32 pop(void) {
    return sp > 0 ? stack[--sp] : 0;
}

/* ── exported API ── */

/* flux_init — zero out VM state */
__attribute__((visibility("default")))
void flux_init(void) {
    for (int i = 0; i < STACK_SIZE; i++) stack[i] = 0;
    sp  = 0;
    gas = 0;
}

/*
 * flux_check — run bytecode on a single input value.
 *   bytecode_ptr: pointer into linear memory
 *   bc_len:       length of bytecode in bytes
 *   input:        the value to validate
 *   returns: 0 = pass, 1 = fail (assertion), 2 = gas exhausted, 3 = stack overflow
 */
__attribute__((visibility("default")))
i32 flux_check(u8 *bytecode_ptr, u32 bc_len, i32 input) {
    sp  = 0;
    gas = 0;
    push(input);                       /* seed stack with the input value */

    u32 pc = 0;
    while (pc < bc_len) {
        if (++gas >= MAX_GAS) return 2; /* gas exhausted */

        u8 op = bytecode_ptr[pc++];

        switch (op) {
        case OP_HALT:
            /* top of stack is the result: 0 = pass */
            return (pop() == 0) ? 0 : 1;

        case OP_ASSERT: {
            i32 cond = pop();
            if (!cond) return 1;        /* assertion failed */
            break;
        }

        case OP_RANGE: {
            i32 hi   = pop();
            i32 lo   = pop();
            i32 val  = pop();
            push((val >= lo && val <= hi) ? 1 : 0);
            break;
        }

        case OP_BOOL_AND: {
            i32 b = pop();
            i32 a = pop();
            push((a && b) ? 1 : 0);
            break;
        }

        case OP_BOOL_OR: {
            i32 b = pop();
            i32 a = pop();
            push((a || b) ? 1 : 0);
            break;
        }

        case OP_DUP: {
            if (sp == 0) return 3;
            i32 v = stack[sp - 1];
            if (sp >= STACK_SIZE) return 3;
            push(v);
            break;
        }

        case OP_SWAP: {
            if (sp < 2) return 3;
            i32 tmp       = stack[sp - 1];
            stack[sp - 1] = stack[sp - 2];
            stack[sp - 2] = tmp;
            break;
        }

        default:
            return 3;                   /* unknown opcode */
        }
    }

    /* fell off the end without HALT — treat as pass if stack top == 0 */
    return (sp > 0 && stack[0] != 0) ? 1 : 0;
}

/*
 * flux_batch — run the same bytecode on an array of inputs.
 *   bytecode_ptr: pointer to bytecode
 *   bc_len:       bytecode length
 *   inputs_ptr:   pointer to i32[n] input array
 *   results_ptr:  pointer to i32[n] output array (0/1/2/3 per element)
 *   n:            number of inputs
 *   max_gas:      per-element gas limit
 *   returns: 0 on success, 1 if any element failed
 */
__attribute__((visibility("default")))
i32 flux_batch(u8 *bytecode_ptr, u32 bc_len, i32 *inputs_ptr,
               i32 *results_ptr, u32 n, u16 max_gas_param) {
    u16 saved_max = MAX_GAS;            /* we'd use a variable but keep it simple */
    i32 any_fail = 0;

    for (u32 i = 0; i < n; i++) {
        sp  = 0;
        gas = 0;
        push(inputs_ptr[i]);

        u32 pc = 0;
        i32 result = 0;                 /* default: pass */

        while (pc < bc_len) {
            if (++gas >= max_gas_param) { result = 2; break; }

            u8 op = bytecode_ptr[pc++];
            switch (op) {
            case OP_HALT:
                result = (pop() == 0) ? 0 : 1;
                goto done;

            case OP_ASSERT: {
                i32 cond = pop();
                if (!cond) { result = 1; goto done; }
                break;
            }

            case OP_RANGE: {
                i32 hi  = pop();
                i32 lo  = pop();
                i32 val = pop();
                push((val >= lo && val <= hi) ? 1 : 0);
                break;
            }

            case OP_BOOL_AND: {
                i32 b = pop(); i32 a = pop();
                push((a && b) ? 1 : 0);
                break;
            }

            case OP_BOOL_OR: {
                i32 b = pop(); i32 a = pop();
                push((a || b) ? 1 : 0);
                break;
            }

            case OP_DUP: {
                if (sp == 0 || sp >= STACK_SIZE) { result = 3; goto done; }
                push(stack[sp - 1]);
                break;
            }

            case OP_SWAP: {
                if (sp < 2) { result = 3; goto done; }
                i32 tmp       = stack[sp - 1];
                stack[sp - 1] = stack[sp - 2];
                stack[sp - 2] = tmp;
                break;
            }

            default:
                result = 3;
                goto done;
            }
        }
done:
        results_ptr[i] = result;
        if (result != 0) any_fail = 1;
    }

    return any_fail;
}
